spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -886,7 +886,14 @@ else if (methodName.equals("close")) {
 
 				// If we get here, we're supposed to shut down.
 				physicalClose();
-				releasePermit();
+				/*
+				 *  Only release a permit if this is a normal close; if the channel is
+				 *  in the list, it means we're closing a cached channel (for which a permit
+				 *  has already been released).
+				 */
+				if (!this.channelList.contains(proxy)) {
+					releasePermit();
+				}
```

Comment:
I commented out this `if` altogether and `testAll` passes.
Even that new `RabbitReconnectProblemTests` passes for me after restarting RabbitMQ.

Or I am missing anything, or we really don't need this logic.
Just because we go to the `channelList` only if `this.channelList.size() < getChannelCacheSize()`.
See the previous `if` in this method:
```
if (!RabbitUtils.isPhysicalCloseRequired() &&
	(this.channelList.size() < getChannelCacheSize()
		|| this.channelList.contains(proxy))) {
	logicalClose((ChannelProxy) proxy);
	// Remain open in the channel list.
	releasePermit();
	return null;
}
```